### PR TITLE
Metrics Aggregation - Improve throughput by 10x

### PR DESCRIPTION
**Pull request by cijothomas**
Part 1 of https://github.com/open-telemetry/opentelemetry-rust/issues/1740

Modifies Sum aggregation (used by Counter/UpDownCounters), to have less contention by using `RwLock` instead of `Mutex` to access `HashMap` of values. Updates (the hot path) now only need `read()` lock as it leverages interior mutability to update the underlying value. This effectively makes the `HashMap` read-heavy, and only need `read()` locks in hot path, **significantly reducing contention**, and thereby boosting throughput.

Perf numbers from metrics stress test confirms the above: we jump from **3 M/sec to 35 M/sec, i.e 10X jump!** 
Criterion based benchmarks uses single thread and hence, won't show contention. They are not expected to change at all with the changes in this PR - and results show no change.

This PR focused on throughput only, and the next set of PRs (which require more refactoring), will boost the benchmarks as well significantly.

### CODE REVIEW

**Comment by lalitb** [comment]
Nice work! updating the value atomically under read lock seems to have boosted the throughout a lot :) Changes look good on quick scan, will review it thoroughly. and though we can get some more juice with #1564 , we are hopefully good for now with this piece of code.

**Comment by cijothomas** [comment]
> Nice work! updating the value atomically under read lock seems to have boosted the throughout a lot :) Changes look good on quick scan, will review it thoroughly. and though we can get some more juice with #1564 , we are hopefully good for now with these changes.

Once contention is avoided (this PR), we are unlikely to gain _hot path_ perf by just sharding alone. But that'll surely help ease the spikes when collect() thread runs, as we can do locks on smaller section instead of whole. We can revisit the sharding logics from 1564.

File: opentelemetry-sdk/src/metrics/internal/sum.rs [hunk 1]
```diff
@@ -1,8 +1,9 @@
+// use std::f32::consts::E;
```
Comment (TommyCpp) [inline opentelemetry-sdk/src/metrics/internal/sum.rs:1]: Is it needed

**Review by hdost — APPROVED** [review]

**Review by lalitb — APPROVED** [review]

**Review by utpilla — APPROVED** [review]

File: opentelemetry-sdk/src/metrics/internal/sum.rs (hunk 1 above)
Comment (cijothomas) [reply opentelemetry-sdk/src/metrics/internal/sum.rs:1]: No! Removed.

**Comment by cijothomas** [comment]
> > Nice work! updating the value atomically under read lock seems to have boosted the throughout a lot :) Changes look good on quick scan, will review it thoroughly. and though we can get some more juice with #1564 , we are hopefully good for now with these changes.
> 
> Once contention is avoided (this PR), we are unlikely to gain _hot path_ perf by just sharding alone. But that'll surely help ease the spikes when collect() thread runs, as we can do locks on smaller section instead of whole. We can revisit the sharding logics from 1564.

Correction. Sharding still helps updates() compete less with other updates() that need to insert a new KVP combination.

**Pull request merged**